crates/bvs-guardrail/src/contract.rs

```diff
@@ -0,0 +1,1453 @@
+            |id| match id {
+                Some(_) => Err(ContractError::ProposalAlreadyExists {}),
+                None => Ok(proposal_id),
+            },
+        )?;
+
+        // TODO: these events are different from cw3 specs to be more inlined with our other contracts
+        Ok(Response::new().add_event(
+            Event::new("propose")
+                .add_attribute("sender", info.sender)
+                .add_attribute("proposal_id", proposal_id.to_string())
+                .add_attribute("slash_request_id", slash_request_id.to_string())
+                .add_attribute("status", format!("{:?}", prop.status)),
+        ))
+    }
+
+    pub fn vote(
+        deps: DepsMut,
+        env: Env,
+        info: MessageInfo,
+        slash_request_id: SlashingRequestId,
+        vote: Vote, // TODO: check if we want to remove abstain and veto votes
+    ) -> Result<Response, ContractError> {
+        // get proposal id
+        let proposal_id = SLASHING_REQUEST_TO_PROPOSAL
+            .may_load(deps.storage, slash_request_id.clone())?
+            .ok_or(ContractError::ProposalNotFound {})?;
+
+        // ensure proposal exists and can be voted on
+        let mut prop = PROPOSALS.load(deps.storage, proposal_id)?;
+        // Allow voting on Passed and Rejected proposals too,
+        if ![Status::Open, Status::Passed, Status::Rejected].contains(&prop.status) {
+            return Err(ContractError::NotOpen {});
+        }
+        // if they are not expired
+        if prop.expires.is_expired(&env.block) {
+            return Err(ContractError::Expired {});
+        }
+
+        // get voter's weight only if weight is >= 1 at start of proposal
+        let vote_power = get_voting_power(deps.storage, &info.sender, prop.start_height)?
+            .ok_or(ContractError::Unauthorized {})?;
+
+        // cast vote if no vote previously cast
+        BALLOTS.update(deps.storage, (proposal_id, &info.sender), |bal| match bal {
+            Some(_) => Err(ContractError::AlreadyVoted {}),
+            None => Ok(Ballot {
+                weight: vote_power,
+                vote,
+            }),
+        })?;
+
+        // update vote tally
+        prop.votes.add_vote(vote, vote_power);
+        prop.update_status(&env.block);
+        PROPOSALS.save(deps.storage, proposal_id, &prop)?;
+
+        Ok(Response::new().add_event(
+            Event::new("vote")
+                .add_attribute("sender", info.sender)
+                .add_attribute("proposal_id", proposal_id.to_string())
+                .add_attribute("slash_request_id", slash_request_id.to_string())
+                .add_attribute("status", format!("{:?}", prop.status))
+                .add_attribute("vote", format!("{:?}", vote)),
+        ))
+    }
+
+    pub fn close(
+        deps: DepsMut,
+        env: Env,
+        info: MessageInfo,
+        slash_request_id: SlashingRequestId,
+    ) -> Result<Response, ContractError> {
+        // anyone can trigger this
```

Comment:
```suggestion
    /// Anyone can close
    pub fn close(
        deps: DepsMut,
        env: Env,
        info: MessageInfo,
        slash_request_id: SlashingRequestId,
    ) -> Result<Response, ContractError> {
```

Maybe move up the comment.